### PR TITLE
#184637134 feat:user-login 

### PR DESCRIPTION
**Pull request by nkurunziza1**
**Description**:

This pull request adds login functionality to the project. Users can now securely log in to their accounts using their email and password. The login feature has been implemented according to industry best practices for security and user experience.

**Testing**:

The login functionality has been tested locally with multiple test accounts to ensure that users can successfully log in and that any errors or issues are properly handled.

**Checklist**:

The code has been thoroughly tested locally and all tests pass.
The login feature has been implemented according to industry best practices.
Security measures, such as encryption of user passwords, have been properly implemented.

### CODE REVIEW

**Review by istafrica-dennw — CHANGES_REQUESTED** [review]
1. you commits are named wrong
2. you have not added reviewers

**Comment by mjeanbosco19** [comment]
@nkurunziza1 Your tests failed on CircleCI, try to work on it

File: src/controllers/user-login-controller.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        token: token,
+      });
+    } catch (error) {
+      return next(error);
+    }
+  })(req, res, next);
+};
+
+const protectedRouteHandler = async (req, res, user) => {
+  res.json({ message: `welcome, ${user.firstname} ` });
```
Comment (Abdulkeza) [inline src/controllers/user-login-controller.js]: what will happen if an invalid user passed in this func?
it's better to catch the errors that might happen.

`const protectedRouteHandler = async (req, res, user) => {
  try {
    res.json({ message: `welcome, ${user.firstname} ` });
  } catch (error) {
    return res.status(500).json({ message: 'Internal server error' });
  }
};`

**Review by Abdulkeza — CHANGES_REQUESTED** [review]
@nkurunziza1 , why do wee need **user.controller.js** and **user-login-controller.js**?

Please can we have either _user.controller.js*_ or _auth.controller.js_*

also user-login.routes.js and user.routes.js ??

File: src/routes/api/user-login.routes.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+
+import { Router } from 'express';
+import { requireAuth } from '../../middlewares/requireAuth.js';
+import { loginUser, protectedRouteHandler } from '../../controllers/user-login-controller.js';
+
+const LoginRouter = Router();
+
+LoginRouter.post('/users/login', loginUser);
```
Comment (leandreAlly) [inline src/routes/api/user-login.routes.js]: you're not following the our folder structure convention, we suppose to only have single routes for users.
why did you added new file?

File: src/controllers/user-login-controller.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import passport from 'passport';
```
Comment (leandreAlly) [inline src/controllers/user-login-controller.js]: also this controller must be binded into user.controller.js

**Review by leandreAlly — CHANGES_REQUESTED** [review]

File: src/routes/api/user.routes.js [hunk 4]
```diff
@@ -1,9 +1,12 @@
 import { Router } from "express";
-import { registerUser } from "../../controllers/user.controller";
+import { registerUser, loginUser, protectedRouteHandler} from "../../controllers/user.controller";
+import { requireAuth } from '../../middlewares/requireAuth.js';
 import verifyUser from "../../middlewares/verifyUser";
 import signupValidation from "../../ validations/signup.validation";
 const route = Router();
 
 route.post("/signup", signupValidation, verifyUser, registerUser);
+route.post('/users/login', loginUser);
```
Comment (ba-lambert) [inline src/routes/api/user.routes.js]: Check this endpoint mhn

**Comment by ba-lambert** [comment]
And also your test coverage  is still under 81

**Comment by mjeanbosco19** [comment]
If possible squash commits to easy rebase issue

**Pull request merged**